core/src/main/java/org/infinispan/conflict/resolution/StateReceiverImpl.java

```diff
@@ -0,0 +1,134 @@
+                    CommandsFactory commandsFactory,
+                    Configuration configuration,
+                    DataContainer<K, V> dataContainer,
+                    LocalTopologyManager localTopologyManager,
+                    RpcManager rpcManager) {
+      this.cache = cache;
+      this.commandsFactory = commandsFactory;
+      this.configuration = configuration;
+      this.dataContainer = dataContainer;
+      this.localTopologyManager = localTopologyManager;
+      this.rpcManager = rpcManager;
+
+      this.transferTimeout = configuration.clustering().stateTransfer().timeout();
+   }
+
+   @Override
+   public CompletableFuture<List<Map<Address, CacheEntry<K, V>>>> getAllReplicasForSegment(int segmentId) {
+      Set<Integer> segments = new HashSet<>();
+      segments.add(segmentId);
+
+      CacheTopology cacheTopology = localTopologyManager.getCacheTopology(cache.getName());
+      ConsistentHash hash = cacheTopology.getWriteConsistentHash();
+      List<Address> replicas = hash.locateOwnersForSegment(segmentId);
+
+      synchronized (keyReplicaMap) {
+         for (Address replica : replicas) {
+            if (replica.equals(rpcManager.getAddress())) {
+               dataContainer.forEach(entry -> {
+                  K key = entry.getKey();
+                  if (hash.getSegment(key) == segmentId) {
+                     addKeyToReplicaMap(replica, entry);
+                  }
+               });
+            } else {
+               InboundTransferTask inboundTransferTask = new InboundTransferTask(segments, replica, cacheTopology.getTopologyId(),
+                     rpcManager, commandsFactory, transferTimeout, cache.getName(), false);
+               transferTaskMap.put(replica, inboundTransferTask);
+               inboundTransferTask.requestSegments();
+            }
+         }
+      }
+
+      return transferFuture.thenApply(v -> getAndClearKeyReplicaMap());
+   }
+
+   private List<Map<Address, CacheEntry<K, V>>> getAndClearKeyReplicaMap() {
+      transferFuture = new CompletableFuture<>();
+      synchronized (keyReplicaMap) {
+         List<Map<Address, CacheEntry<K, V>>> retVal = keyReplicaMap.entrySet().stream()
+               .map(Map.Entry::getValue)
+               .collect(Collectors.toList());
+         keyReplicaMap.clear();
+         return retVal;
+      }
+   }
+
+   @Override
+   public void receiveState(Address sender, Collection<StateChunk> stateChunks) {
+      synchronized (keyReplicaMap) {
+         InboundTransferTask transferTask = transferTaskMap.get(sender);
+         boolean lastChunkReceived = false;
+         for (StateChunk chunk : stateChunks) {
+            transferTask.onStateReceived(chunk.getSegmentId(), chunk.isLastChunk());
+            chunk.getCacheEntries().forEach(ice -> addKeyToReplicaMap(sender, ice));
+            lastChunkReceived = chunk.isLastChunk();
+         }
+
+         if (lastChunkReceived) {
+            transferFuture.complete(null);
```

Comment:
It seems that this is the only use of transferFuture - could you call `getAndClearKeyReplicaMap` directly?